crates/voicevox_core/src/user_dict/word.rs

```diff
@@ -0,0 +1,222 @@
+            )));
+        }
+        Self::validate_pronunciation(&pronunciation[..])?;
+        let mora_count = Self::calculate_mora_count(&pronunciation[..], accent_type)?;
+        Ok(Self {
+            surface: Self::to_zenkaku(&surface[..]),
+            pronunciation,
+            accent_type,
+            word_type,
+            priority,
+            mora_count,
+        })
+    }
+
+    /// カタカナの文字列が発音として有効かどうかを判定する。
+    fn validate_pronunciation(pronunciation: &str) -> Result<()> {
+        // 元実装：https://github.com/VOICEVOX/voicevox_engine/blob/39747666aa0895699e188f3fd03a0f448c9cf746/voicevox_engine/model.py#L190-L210
+        if !PRONUNCIATION_REGEX.is_match(pronunciation) {
+            return Err(Error::InvalidWord(
+                "発音は有効なカタカナである必要があります。".to_string(),
+            ));
+        }
+        let sutegana = ['ァ', 'ィ', 'ゥ', 'ェ', 'ォ', 'ャ', 'ュ', 'ョ', 'ヮ', 'ッ'];
+
+        let pronunciation_chars = pronunciation.chars().collect::<Vec<_>>();
+
+        for i in 0..pronunciation_chars.len() {
+            // 「キャット」のように、捨て仮名が連続する可能性が考えられるので、
+            // 「ッ」に関しては「ッ」そのものが連続している場合と、「ッ」の後にほかの捨て仮名が連続する場合のみ無効とする
+            if sutegana.contains(pronunciation_chars.get(i).unwrap())
+                && i < pronunciation_chars.len() - 1
+                && (sutegana[..sutegana.len() - 1]
+                    .contains(pronunciation_chars.get(i + 1).unwrap())
+                    || (pronunciation_chars.get(i).unwrap() == &'ッ'
+                        && sutegana.contains(pronunciation_chars.get(i + 1).unwrap())))
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(捨て仮名の連続)".to_string(),
+                ));
+            }
+
+            if pronunciation_chars.get(i).unwrap() == &'ヮ'
+                && i != 0
+                && !['ク', 'グ'].contains(pronunciation_chars.get(i - 1).unwrap())
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(「くゎ」「ぐゎ」以外の「ゎ」の使用)".to_string(),
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// カタカナの発音からモーラ数を計算する。
+    fn calculate_mora_count(pronunciation: &str, accent_type: usize) -> Result<usize> {
+        // 元実装：https://github.com/VOICEVOX/voicevox_engine/blob/39747666aa0895699e188f3fd03a0f448c9cf746/voicevox_engine/model.py#L212-L236
+        let mora_count = MORA_REGEX.find_iter(pronunciation).count();
+
+        if accent_type > mora_count {
+            return Err(Error::InvalidWord(format!(
+                "誤ったアクセント型です({})。 expect: 0 <= accent_type <= {}",
+                accent_type, mora_count
+            )));
+        }
+
+        Ok(mora_count)
+    }
+
+    /// ASCII文字を全角文字に変換する。
+    fn to_zenkaku(surface: &str) -> String {
+        // 元実装：https://github.com/VOICEVOX/voicevox/blob/69898f5dd001d28d4de355a25766acb0e0833ec2/src/components/DictionaryManageDialog.vue#L379-L387
+        SPACE_REGEX
+            // " "などの目に見えない文字をまとめて全角スペース(0x3000)に置き換える
```

Comment:
`\p{Z}`をすべて全角スペースに置き換えるのはインターフェイスとしての挙動の話なので、docstring側に持って来るのはどうでしょうか。

また、「ASCII文字」だと正確ではないと思うので、次の形を提案します。
(「"!"から"~"までの〜」は元実装から取ってきました)

```suggestion
    /// 一部の種類の文字を、全角文字に置き換える。
    ///
    /// 具体的には
    /// - "!"から"~"までの範囲の文字(数字やアルファベット)は、対応する全角文字に
    /// - " "などの目に見えない文字は、まとめて全角スペース(0x3000)に
    /// 変換する。
    fn to_zenkaku(surface: &str) -> String {
        // 元実装：https://github.com/VOICEVOX/voicevox/blob/69898f5dd001d28d4de355a25766acb0e0833ec2/src/components/DictionaryManageDialog.vue#L379-L387
        SPACE_REGEX
```